File: chapter06/mrtr-client/typescript/src/native.ts
// The SDK-native MRTR path: no hand-written loop.
//
// In v2, `client.callTool()` fulfils `input_required` results AUTOMATICALLY
// by default (ClientOptions.inputRequired.autoFulfill defaults to true): it
// dispatches each embedded request to the handler registered via
// `setRequestHandler('elicitation/create', ...)`, then retries the original
// call with the collected inputResponses and a byte-exact requestState echo,
// on a fresh request id, up to `maxRounds` (default 10) rounds. The
// interactive rounds happen inside the call — callTool resolves with the
// final result. `maxTotalTimeout` bounds the whole flow, and an abort signal
// cancels it, so the chapter's retry budget / timeout / cancel obligations
// map directly onto SDK options here.
//
// What the SDK does NOT do: render the form, apply policy, or validate the
// user's content against requestedSchema. Those stay in your handler —
// the same handleInputRequest the manual loop uses.
//
// Run: node dist/native.js  (same MRTR_ANSWERS / MRTR_POLICY env as the manual entry)

import { Client } from '@modelcontextprotocol/client';
import { StdioClientTransport } from '@modelcontextprotocol/client/stdio';
import { handleInputRequest } from './input-handler.js';

const client = new Client(
  { name: 'mrtr-native-client', version: '0.1.0' },
  {
    versionNegotiation: { mode: 'auto' },
    // Required for the handler registration below AND so the server sees the
    // elicitation capability in the per-request _meta envelope.
    capabilities: { elicitation: {} },
    inputRequired: { maxRounds: 5 }, // autoFulfill: true is the default
  },
);

// The driver dispatches embedded elicitation requests to this handler.
client.setRequestHandler('elicitation/create', async (request) => handleInputRequest(request));

await client.connect(new StdioClientTransport({ command: 'node', args: ['dist/demo-server.js'] }));

console.log('-> tools/call book_meeting (SDK drives the MRTR rounds)');
try {
  const result = await client.callTool(
    { name: 'book_meeting', arguments: { room: '4B' } },
    {
      maxTotalTimeout: 120_000,
      // The driver reports each fulfilment round as progress.
      onprogress: (p) => console.log(`  [driver] ${p.message}`),
    },
  );
  const text = result.content.find((c) => c.type === 'text');
  console.log(`<- final result: ${text?.type === 'text' ? text.text : JSON.stringify(result.content)}`);
} catch (error) {
  console.log(`<- failed: ${error instanceof Error ? error.message : String(error)}`);
  process.exitCode = 1;
} finally {
  await client.close();
}

File: chapter06/mrtr-client/python/native.py
# The SDK-native MRTR path: no hand-written loop.
#
# In mcp 2.0.0, `client.call_tool()` (and `get_prompt` / `read_resource`)
# resolves `InputRequiredResult`s AUTOMATICALLY: each embedded request is
# dispatched to the matching callback (`elicitation_callback` here; sampling
# and roots have their own), all requests in one round run concurrently, and
# the call is retried with the collected `input_responses` and a byte-exact
# `request_state` echo, on a fresh request id, up to `input_required_max_rounds`
# (default 10, `InputRequiredRoundsExceededError` beyond it). A result carrying
# only `request_state` — no requests — is retried after a short exponential
# backoff (50 ms doubling to a 250 ms cap): the load-shedding leg. The
# interactive rounds happen inside the call — `call_tool` returns the final
# result.
#
# What the SDK does NOT do: render the form, apply policy, or validate the
# user's content against requested_schema. Those stay in your callback —
# the same handle_elicitation the manual loop's handler uses.
#
# Run: uv run native.py  (same MRTR_ANSWERS / MRTR_POLICY env as the manual entry)

import os
import sys

import anyio
from mcp.client import Client
from mcp.client.stdio import StdioServerParameters, stdio_client
from mcp_types import TextContent

from input_handler import handle_elicitation


async def main() -> int:
    tool_name = sys.argv[1] if len(sys.argv) > 1 else "book_meeting"
    here = os.path.dirname(os.path.abspath(__file__))
    transport = stdio_client(
        StdioServerParameters(command=sys.executable, args=[os.path.join(here, "demo_server.py")])
    )
    client = Client(
        transport,
        # Registering the callback also advertises the elicitation capability.
        elicitation_callback=lambda context, params: handle_elicitation(params),
        input_required_max_rounds=int(os.environ.get("MRTR_MAX_ROUNDS", "5")),
    )

    async with client:
        print(f"-> tools/call {tool_name} (SDK drives the MRTR rounds; negotiated {client.protocol_version})")
        arguments = {"room": "4B"} if tool_name == "book_meeting" else {}
        try:
            result = await client.call_tool(tool_name, arguments)
        except Exception as error:  # includes InputRequiredRoundsExceededError
            print(f"<- failed: {error}")
            return 1
        text = next((c.text for c in result.content if isinstance(c, TextContent)), None)
        print(f"<- final result: {text if text is not None else result.content}")
        return 0


if __name__ == "__main__":
    try:
        sys.exit(anyio.run(main))
    except KeyboardInterrupt:
        print("<- failed: tool call cancelled by the user")
        sys.exit(1)
